docs/usermanual-clusters.xml

```diff
@@ -5,306 +5,509 @@
-    One last note about levels 0 and 1. We currently don't allow a
-    <literal>MultipleSubst</literal> lookup to replace a glyph with zero
-    glyphs (i.e., to delete a glyph). But in some other situations,
-    glyphs can be deleted. In those cases, if the glyph being deleted is
-    the last glyph of its cluster, we make sure to merge the cluster
-    with a neighboring cluster.
-  </para>
-  <para>
-    This is, primarily, to make sure that the starting cluster of the
-    text always has the cluster index pointing to the start of the text
-    for the run; more than one client currently relies on this
-    guarantee.
-  </para>
-  <para>
-    Incidentally, Apple's CoreText does something else to maintain the
-    same promise: it inserts a glyph with id 65535 at the beginning of
-    the glyph string if the glyph corresponding to the first character
-    in the run was deleted. HarfBuzz might do something similar in the
-    future.
-  </para>
-</sect1>
-<sect1 id="level-2">
-  <title>Level 2</title>
-  <para>
-    Level 2 is a different beast from levels 0 and 1. It is simple to
-    describe, but hard to make sense of. It simply doesn't do any
-    cluster merging whatsoever. When things ligate or otherwise multiple
-    glyphs turn into one, the cluster value of the first glyph is
-    retained.
-  </para>
-  <para>
-    Here are a few examples of why processing cluster values produced at
-    this level might be tricky:
-  </para>
-  <sect2 id="ligatures-with-combining-marks">
-    <title>Ligatures with combining marks</title>
-    <para>
-      Imagine capital letters are bases and lower case letters are
-      combining marks. With an input sequence like this:
+  <section id="clusters">
+    <title>Clusters</title>
+    <para>
+      In text shaping, a <emphasis>cluster</emphasis> is a sequence of
+      characters that needs to be treated as a single, indivisible
+      unit.
     </para>
-    <programlisting>
-  A,a,B,b,C,c
-  0,1,2,3,4,5
-</programlisting>
     <para>
-      if <literal>A,B,C</literal> ligate, then here are the cluster
-      values one would get under the various levels:
+      During the shaping process, some shaping operations may
+      merge adjacent characters (for example, when two code points form
+      a ligature and are replaced by a single glyph) or split one
+      character into several (for example, when performing the Unicode
+      canonical decomposition of a code point).
     </para>
     <para>
-      level 0:
+      HarfBuzz tracks clusters independently from how these
+      shaping operations alter the individual glyphs that comprise the
+      output HarfBuzz returns in a buffer. Consequently,
+      a client program using HarfBuzz can utilize the cluster
+      information to implement features such as:
+    </para>
+    <itemizedlist>
+      <listitem>
+	<para>
+	  Correctly positioning the cursor between two characters that
+	  have combined into a single glyph by forming a ligature.
```

Comment:
Okay, but if many-to-one substitutions like ligation didn't exist, then it would always be a trivial problem to solve, wouldn't it?

Reply:
It is trivial only when you have linear one-to-one mapping, so that character position matches glyph position, many-to-one, one-to-many, and reordering all break that assumption and make the mapping none-trivial.